File: Claude/Python/2024/Day25_Solution_2024.py
with open("../../../Codex/Python/2024/Day25_Input.txt") as f:
    text = f.read()

locks, keys = [], []

for block in text.strip().split("\n\n"):
    rows = block.splitlines()
    if rows[0] == "#####":  # lock: top filled
        heights = [sum(rows[r][c] == '#' for r in range(1, 7)) for c in range(5)]
        locks.append(heights)
    else:  # key: bottom filled
        heights = [sum(rows[r][c] == '#' for r in range(0, 6)) for c in range(5)]
        keys.append(heights)

print(sum(
    all(l + k <= 5 for l, k in zip(lock, key))
    for lock in locks
    for key in keys
))

File: Claude/Rust/2024/Day25_Solution_2024.rs
use std::time::Instant;
use std::fs;

fn main() {
    let start = Instant::now();
    let input = fs::read_to_string("Day25_Input.txt").expect("Failed to read input file");

    let mut locks: Vec<[u8; 5]> = Vec::new();
    let mut keys: Vec<[u8; 5]> = Vec::new();

    for block in input.split("\n\n") {
        let rows: Vec<&str> = block.lines().filter(|l| !l.is_empty()).collect();
        if rows.len() != 7 {
            continue;
        }
        let is_lock = rows[0] == "#####";
        let mut heights = [0u8; 5];
        // For locks: count '#' in rows 1..=6 per column (top row is all '#', skip it)
        // For keys:  count '#' in rows 0..=5 per column (bottom row is all '#', skip it)
        let range = if is_lock { 1..7 } else { 0..6 };
        for r in range {
            for (c, ch) in rows[r].chars().enumerate().take(5) {
                if ch == '#' {
                    heights[c] += 1;
                }
            }
        }
        if is_lock {
            locks.push(heights);
        } else {
            keys.push(heights);
        }
    }

    let mut count = 0usize;
    for lock in &locks {
        for key in &keys {
            if (0..5).all(|i| lock[i] + key[i] <= 5) {
                count += 1;
            }
        }
    }

    println!("{}", count);

    println!("Runtime: {:.6} seconds", start.elapsed().as_secs_f64());
}
